enhancements/machine-config/onprem-namespaces.md

```diff
@@ -0,0 +1,134 @@
+---
+title: namespaces-for-on-prem-networking-services
+authors:
+  - "@cybertron"
+reviewers:
+  - "@smarterclayton"
+  - "@celebdor"
+  - "@mandre"
+  - "@rgolangh"
+  - "@patrickdillon"
+  - "@kikisdeliveryservice"
+approvers:
+  - Someone from MCO
+  - Someone from architecture
+creation-date: 2020-07-24
+last-updated: 2020-07-27
+status: implementable
+see-also:
+replaces:
+superseded-by:
+---
+
+# Namespaces for On-Prem Networking Services
+
+## Release Signoff Checklist
+
+- [ ] Enhancement is `implementable`
+- [ ] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)
+
+## Summary
+
+When support for on-premise deployments on baremetal, OpenStack, OVirt, and
+VSphere was added, some additional networking services were needed to provide
+functionality that would come from cloud services on other platforms. These
+services were added in namespaces called openshift-kni-infra,
+openshift-openstack-infra, etc. There are two problems with this: 1) The
+namespaces are created in all deployments, including on platforms other
+than the one that uses the namespace and 2) The openshift-x-infra name scheme
+does not match the existing pattern for OpenShift namespaces.
+
+## Motivation
+
+The existence of namespaces not related to a given deployment is confusing to
+users. Additionally, the inconsistent naming of the namespaces is a further
+possible point of confusion.
+
+### Goals
+
+- Only create namespaces for networking services on deployments that need them.
+- Use names that are consistent with the rest of OpenShift.
+
+### Non-Goals
+
+## Proposal
+
+Instead of creating the namespaces at install time as is currently done, add
+them to the list of
+[resources to be synced](https://github.com/openshift/machine-config-operator/blob/master/pkg/operator/sync.go)
+in the machine-config-operator. This will allow us to add logic that considers
+the platform before creating them.
+
+For the naming, we propose `openshift-hosted-network-services` for all
+platforms that use these components. There is no need for a platform-specific
+name since only one platform can exist at a time anyway.
+
+### User Stories [optional]
+
+### Implementation Details/Notes/Constraints [optional]
+
+The services in question provide things like internal DNS resolution between
```

Comment:
This paragraph seems out of place in this section of implementation as it explains what these internal services are?

what we should include here is,

- how we create the resources on bootstrap host
who(binary) is going to be responsible for this
what inputs we use and from where,
where we write files as output
- how we create the necessary resources on bootstrap host for static pods that will run on the control plane
who(binary) is going to be responsible for this
what inputs we use and from where,
where we write files as output

- how do we migrate existing clusters to move from previous namespaces to this new namespace
who(binary) is going to be responsible for this
What exactly we will be doing, and when (during the upgrade time).
What are our failure cases and how we intend to get these fix them.
How are we going to make sure the migration is zero down time.

Reply:
Yeah, this was just trying to provide context for what we're talking about here. I can move it somewhere else and try to address those points.

Reply:
I don't understand why we would want the MCO to handle this.  CVO creates the manifests today, it should be in charge of filtering what gets applied and what doesn't.  We already do that today based on annotations: https://github.com/openshift/cluster-version-operator/blob/master/pkg/payload/payload.go#L183

```
exclude.release.openshift.io/internal-openshift-hosted: "true"
```

If there are certain operators we know will only be included on a particular infrastructure type, we could make an include-when type annotation.